File: app/src/main/java/org/mesibo/messenger/ShowProfileActivity.java
/******************************************************************************
* By accessing or copying this work, you agree to comply with the following   *
* terms:                                                                      *
*                                                                             *
* Copyright (c) 2019-2023 mesibo                                              *
* https://mesibo.com                                                          *
* All rights reserved.                                                        *
*                                                                             *
* Redistribution is not permitted. Use of this software is subject to the     *
* conditions specified at https://mesibo.com . When using the source code,    *
* maintain the copyright notice, conditions, disclaimer, and  links to mesibo * 
* website, documentation and the source code repository.                      *
*                                                                             *
* Do not use the name of mesibo or its contributors to endorse products from  *
* this software without prior written permission.                             *
*                                                                             *
* This software is provided "as is" without warranties. mesibo and its        *
* contributors are not liable for any damages arising from its use.           *
*                                                                             *
* Documentation: https://mesibo.com/documentation/                            *
*                                                                             *
* Source Code Repository: https://github.com/mesibo/                          *
*******************************************************************************/

package org.mesibo.messenger;


import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import com.google.android.material.appbar.AppBarLayout;
import com.google.android.material.appbar.CollapsingToolbarLayout;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import android.view.Menu;
import android.view.View;
import android.view.animation.AlphaAnimation;
import android.widget.TextView;

import com.mesibo.api.Mesibo;
import com.mesibo.api.MesiboDateTime;
import com.mesibo.api.MesiboProfile;

public class ShowProfileActivity extends AppCompatActivity implements ShowProfileFragment.OnFragmentInteractionListener, MesiboProfile.Listener {

    SquareImageView mUsermageView;
    MesiboProfile mUserProfile;
    Toolbar mToolbar;
    AppBarLayout mAppBarLayout;
    CoordinatorLayout mCoordinatorLayout;

    long mGroupId = 0;
    String mPeer = null;
   
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_show_user_profile);

        mToolbar = (Toolbar) findViewById(R.id.up_toolbar);
        setSupportActionBar(mToolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        Bundle args = getIntent().getExtras();
        if (null == args) {
            return;
        }

        mPeer = args.getString("peer");
        mGroupId = args.getLong("groupid");

        mUserProfile = null;

        if (mGroupId > 0) {
            mUserProfile = Mesibo.getProfile(mGroupId);
        } else {
            mUserProfile = Mesibo.getProfile(mPeer);
        }

        mUserProfile.addListener(this);

        mUsermageView = (SquareImageView) findViewById(R.id.up_image_profile);

        Mesibo.addListener(this);

        mUsermageView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                UIManager.launchImageViewer(ShowProfileActivity.this, mUserProfile.getImageOrThumbnailPath());
            }

        });

        TextView userName = (TextView) findViewById(R.id.up_user_name);
        TextView userstatus = (TextView) findViewById(R.id.up_current_status);

        userName.setText(mUserProfile.getName());
        MesiboDateTime lastSeen = mUserProfile.getLastSeen();
        userstatus.setVisibility(View.VISIBLE);
        if(mUserProfile.isOnline()) {
            userstatus.setText("Online");
        }
        else if(null == lastSeen) {
            // never seen or group
            userstatus.setVisibility(View.GONE);
        }
        else {
            String seenStatus = lastSeen.getDateInNaturalLanguage();
            userstatus.setText("Last seen, " + seenStatus);
        }

        CollapsingToolbarLayout collapsingToolbar =
                (CollapsingToolbarLayout) findViewById(R.id.up_collapsing_toolbar);
        collapsingToolbar.setTitle("  ");
        mCoordinatorLayout = (CoordinatorLayout) findViewById(R.id.up_profile_root);
        mAppBarLayout = (AppBarLayout) findViewById(R.id.up_appbar);
        mAppBarLayout.post(new Runnable() {
            @Override
            public void run() {

                setAppBarOffset(-250);
            }
        });

        mAppBarLayout.addOnOffsetChangedListener(new AppBarLayout.OnOffsetChangedListener() {
            @Override
            public void onOffsetChanged(AppBarLayout appBarLayout, int verticalOffset) {
                //measuring for alpha
                float Alpha = ((float) (appBarLayout.getTotalScrollRange() - Math.abs(verticalOffset)) / appBarLayout.getTotalScrollRange());
                if (Alpha > 0.4)
                    Alpha = 1;
                else {
                    Alpha = Alpha + (float) 0.6;
                }
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB)
                    mUsermageView.setAlpha(Alpha);
                else {
                    AlphaAnimation alpha = new AlphaAnimation(Alpha, Alpha);
                    alpha.setDuration(0);
                    alpha.setFillAfter(true);
                    mUsermageView.startAnimation(alpha);
                }


            }
        });

        FragmentManager fragmentManager = getSupportFragmentManager();
        FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
        ShowProfileFragment showUserProfileDetailsFragment = ShowProfileFragment.newInstance(mUserProfile);
        fragmentTransaction.add(R.id.up_fragment, showUserProfileDetailsFragment, "up");
        fragmentTransaction.commit();

    }

    private void setAppBarOffset(int offsetPx) {
        CoordinatorLayout.LayoutParams params = (CoordinatorLayout.LayoutParams) mAppBarLayout.getLayoutParams();
        AppBarLayout.Behavior behavior = (AppBarLayout.Behavior) params.getBehavior();
        behavior.setTopAndBottomOffset(-60 - mAppBarLayout.getTotalScrollRange() / 2);
        //behavior.onNestedPreScroll(mCoordinatorLayout, mAppBarLayout, null, 0, offsetPx, new int[]{0, 0});
    }

    private void setUserPicture() {
        Bitmap b = mUserProfile.getImageOrThumbnail();
        if(null != b) {
            mUsermageView.setImageBitmap(b);
        }
    }

    @Override
    public void MesiboProfile_onUpdate(MesiboProfile userProfile) {

        if(Mesibo.isUiThread()) {
            setUserPicture();
            return;
        }

        new Handler(Looper.getMainLooper()).post(new Runnable() {
            @Override
            public void run() {
                setUserPicture();
            }
        });
    }

    @Override
    public void MesiboProfile_onEndToEndEncryption(MesiboProfile mesiboProfile, int i) {

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        //getMenuInflater().inflate(R.menu.user_profile_menu, menu);
        return true;
    }

    @Override
    public void onFragmentInteraction(Uri uri) {

    }

    @Override
    public void onResume() {
        super.onResume();

        Mesibo.setForegroundContext(this, 0x102, true);

        if(mUserProfile.groupid > 0) {
            TextView userName = (TextView) findViewById(R.id.up_user_name);
            if(null != mUserProfile.getName())
                userName.setText(mUserProfile.getName());
        }

        setUserPicture();
    }

    @Override
    protected void onPause() {
        super.onPause();
        Mesibo.setForegroundContext(this, 0x102, false);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        Mesibo.setForegroundContext(this, 0x102, false);
    }


    @Override
    public void finish() {
        super.finish();
        overridePendingTransition(0, 0);
    }

}

File: app/src/main/java/org/mesibo/messenger/ShowProfileActivity.kt
/** Copyright (c) 2019 Mesibo
 * https://mesibo.com
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the terms and condition mentioned on https://mesibo.com
 * as well as following conditions are met:
 *
 * Redistributions of source code must retain the above copyright notice, this list
 * of conditions, the following disclaimer and links to documentation and source code
 * repository.
 *
 * Redistributions in binary form must reproduce the above copyright notice, this
 * list of conditions and the following disclaimer in the documentation and/or other
 * materials provided with the distribution.
 *
 * Neither the name of Mesibo nor the names of its contributors may be used to endorse
 * or promote products derived from this software without specific prior written
 * permission.
 *
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
 * OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * Documentation
 * https://mesibo.com/documentation/
 *
 * Source Code Repository
 * https://github.com/mesibo/messengerKotlin-app-android
 *
 */

package org.mesibo.messenger


import android.graphics.BitmapFactory
import android.net.Uri
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.support.design.widget.AppBarLayout
import android.support.design.widget.CollapsingToolbarLayout
import android.support.design.widget.CoordinatorLayout
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v7.widget.Toolbar
import android.view.Menu
import android.view.View
import android.view.animation.AlphaAnimation
import android.widget.TextView

import com.mesibo.api.Mesibo

class ShowProfileActivity : AppCompatActivity(), ShowProfileFragment.OnFragmentInteractionListener, Mesibo.FileTransferListener, Mesibo.UserProfileUpdateListener {

    internal lateinit var mUsermageView: SquareImageView
    internal var mUserProfile: Mesibo.UserProfile? = null
    internal lateinit var mToolbar: Toolbar
    internal lateinit var mAppBarLayout: AppBarLayout
    internal lateinit var mCoordinatorLayout: CoordinatorLayout
    internal var mGroupId: Long = 0
    internal var mPeer: String? = null
    private var mProfilePicturePath: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_show_user_profile)
        mToolbar = findViewById<View>(R.id.up_toolbar) as Toolbar
        setSupportActionBar(mToolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        val args = intent.extras ?: return

        mPeer = args.getString("peer")
        mGroupId = args.getLong("groupid")

        mUserProfile = null

        if (mGroupId > 0) {
            mUserProfile = Mesibo.getUserProfile(mGroupId)
        } else {
            mUserProfile = Mesibo.getUserProfile(mPeer)
        }
        mUsermageView = findViewById<View>(R.id.up_image_profile) as SquareImageView

        Mesibo.addListener(this)

        mUsermageView.setOnClickListener { UIManager.launchImageViewer(this@ShowProfileActivity, mProfilePicturePath!!) }

        val userName = findViewById<View>(R.id.up_user_name) as TextView
        val userstatus = findViewById<View>(R.id.up_current_status) as TextView

        userName.text = mUserProfile!!.name
        var lastSeen = Mesibo.getTimestamp() - mUserProfile!!.lastActiveTime
        userstatus.visibility = View.VISIBLE
        if (lastSeen <= 60000) {
            userstatus.text = "Online"
        } else {
            var seenStatus = ""
            lastSeen = lastSeen / 60000 //miutes
            if (mUserProfile!!.groupid > 0 || 0L == mUserProfile!!.lastActiveTime) {
                userstatus.visibility = View.GONE
            } else if (lastSeen >= 2 * 24 * 60) {
                seenStatus = (lastSeen / (24 * 60)).toInt().toString() + " days ago"
            } else if (lastSeen >= 24 * 60) {
                seenStatus = "yesterday"
            } else if (lastSeen >= 120) {
                seenStatus = (lastSeen / 60).toInt().toString() + " hours ago"
            } else if (lastSeen >= 60) {
                seenStatus = "an hour ago"
            } else if (lastSeen >= 2) {
                seenStatus = "$lastSeen minutes ago"
            } else {
                seenStatus = "a few moments before"
            }

            userstatus.text = "Last seen $seenStatus"

            //userstatus.setVisibility(View.GONE);
        }

        val collapsingToolbar = findViewById<CollapsingToolbarLayout>(R.id.up_collapsing_toolbar)
        collapsingToolbar.title = "  "
        mCoordinatorLayout = findViewById(R.id.up_profile_root)
        mAppBarLayout = findViewById(R.id.up_appbar)
        mAppBarLayout.post { setAppBarOffset(-250) }

        mAppBarLayout.addOnOffsetChangedListener(AppBarLayout.OnOffsetChangedListener { appBarLayout, verticalOffset ->
            //measuring for alpha
            var Alpha = (appBarLayout.totalScrollRange - Math.abs(verticalOffset)).toFloat() / appBarLayout.totalScrollRange
            if (Alpha > 0.4)
                Alpha = 1f
            else {
                Alpha = Alpha + 0.6.toFloat()
            }
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB)
                mUsermageView.alpha = Alpha
            else {
                val alpha = AlphaAnimation(Alpha, Alpha)
                alpha.duration = 0
                alpha.fillAfter = true
                mUsermageView.startAnimation(alpha)
            }
        })

        val fragmentManager = supportFragmentManager
        val fragmentTransaction = fragmentManager.beginTransaction()
        val showUserProfileDetailsFragment = ShowProfileFragment.newInstance(mUserProfile!!)
        fragmentTransaction.add(R.id.up_fragment, showUserProfileDetailsFragment, "up")
        fragmentTransaction.commit()
    }

    private fun setAppBarOffset(offsetPx: Int) {
        val params = mAppBarLayout.layoutParams as CoordinatorLayout.LayoutParams
        val behavior = params.behavior as AppBarLayout.Behavior?
        behavior!!.topAndBottomOffset = -60 - mAppBarLayout.totalScrollRange / 2
        //behavior.onNestedPreScroll(mCoordinatorLayout, mAppBarLayout, null, 0, offsetPx, new int[]{0, 0});
    }

    private fun setUserPicture() {
        mProfilePicturePath = Mesibo.getUserProfilePicturePath(mUserProfile, Mesibo.FileInfo.TYPE_AUTO)
        if (null == mProfilePicturePath || !Mesibo.fileExists(mProfilePicturePath))
            return

        val b = BitmapFactory.decodeFile(mProfilePicturePath)
        if (null != b) {
            mUsermageView.setImageBitmap(b)
        }
    }

    override fun Mesibo_onUserProfileUpdated(userProfile: Mesibo.UserProfile, i: Int, refresh: Boolean) {
        if (!refresh)
            return

        if (userProfile !== mUserProfile)
            return

        if (Mesibo.isUiThread()) {
            setUserPicture()
            Mesibo.startUserProfilePictureTransfer(mUserProfile, this)
            return
        }

        Handler(Looper.getMainLooper()).post {
            setUserPicture()
            Mesibo.startUserProfilePictureTransfer(mUserProfile, this@ShowProfileActivity)
        }


    }

    /**
     * @param file
     */
    override fun Mesibo_onFileTransferProgress(file: Mesibo.FileInfo): Boolean {

        if (100 == file.progress)
            setUserPicture()

        return true
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        //getMenuInflater().inflate(R.menu.user_profile_menu, menu);
        return true
    }

    override fun onFragmentInteraction(uri: Uri) {

    }

    public override fun onResume() {
        super.onResume()

        if (!Mesibo.setAppInForeground(this, 0x102, true)) {
            finish()
            return
        }

        if (mUserProfile!!.groupid > 0) {
            val userName = findViewById<View>(R.id.up_user_name) as TextView
            if (null != mUserProfile!!.name)
                userName.text = mUserProfile!!.name
        }

        setUserPicture()
        Mesibo.startUserProfilePictureTransfer(mUserProfile, this)

    }

    override fun onPause() {
        super.onPause()
        Mesibo.setAppInForeground(this, 0x102, false)
    }

    override fun onDestroy() {
        super.onDestroy()
        Mesibo.setAppInForeground(this, 0x102, false)
    }


    override fun finish() {
        super.finish()
        overridePendingTransition(0, 0)
    }


    public override fun onStart() {
        super.onStart()
    }

    public override fun onStop() {
        super.onStop()
    }
}
